slack_today_i_did/parser.py

```diff
@@ -59,50 +80,107 @@ def tokenize(text, known_tokens):
                 func = known_functions[function_name]
+                actual_arg = [Constant(arg_to_function, str)] if len(arg_to_function) > 0 else []
+                return_type = func.__annotations__.get('return', None)
+                args.append(FuncCall(function_name, actual_arg, return_type))
 
-                try:
-                    evaled = func(arg_to_function)
-                    return_type = func.__annotations__.get('return', None)
-                    args.append((evaled, return_type))
-                except Exception as e:
-                    errors.append((function_name, e))
+    action = known_functions[first_function_name]
+    return_type = action.__annotations__.get('return', None)
+    evaluator = functools.partial(evaluate_func_call, known_functions)
 
     return {
-        'action': known_functions[first_function_name],
-        'args': args,
-        'errors': errors
+        'func_call': FuncCall(first_function_name, args, return_type),
+        'evaluate': evaluator,
+    }
+
+
+def evaluate_func_call(known_functions, func_call, default_args=[]) -> FuncResult:
+    all_errors = []
+    action = known_functions[func_call.func_name]
+
+    if is_metafunc(action):
+        args = default_args + [Constant(func_call.args, List[FuncArg])]
+    else:
+        args = default_args + func_call.args
+
+    args_evaluation = evaluate_args(known_functions, args)
+    evaluated_args = args_evaluation['args']
+
+    if len(args_evaluation['errors']) > 0:
+        all_errors.extend(args_evaluation['errors'])
+        return FuncResult(None, None, [], all_errors)
+
+    argument_errors = []
+
+    # TODO: simply copy.deepcopy and pop('return', None) after
+    # https://github.com/python/typing/issues/306 is resolved.
+    # `-> ChannelMessages` blows up if you try to copy.deepcopy the annotations.
+    annotations = dict(action.__annotations__)
+    annotations.pop('return', None)
+    annotations = copy.deepcopy(annotations)
+
+    # check arity mismatch
+    num_keyword_args = len(action.__defaults__) if action.__defaults__ else 0
+    num_positional_args = len(annotations) - num_keyword_args
+    if num_positional_args > len(evaluated_args):
+        argument_errors.append(
+            mismatching_args_messages(action, annotations, evaluated_args)
+        )
+
+    mismatching_types = mismatching_types_messages(
+        action,
+        annotations,
+        evaluated_args
+    )
+
+    if len(mismatching_types) > 0:
+        argument_errors.append(mismatching_types)
+
+    if len(argument_errors) > 0:
+        all_errors.extend(argument_errors)
+        return FuncResult(None, None, [], all_errors)
+
+    try:
+        return FuncResult(action(*evaluated_args), action, evaluated_args, all_errors)
+    except Exception as e:
+        return FuncResult(None, None, [], [exception_error_messages([(func_call.func_name, e)])])
+
+
+def evaluate_args(known_functions, args, default_args=[]):
```

Comment:
:thought_balloon: maybe worth adding type annotations here to explain what these things should be